src/amp-story-player/amp-story-player-iframe-pool.js

```diff
@@ -77,4 +77,63 @@ export class IframePool {
 
     return detachedStoryIdx;
   }
+
+  /**
+   * Clears stored story indices.
+   * @return {!Array<number>}
+   */
+  evictStories() {
+    const evictedStories = this.storyIdsWithIframe_;
+    this.storyIdsWithIframe_ = [];
+    return evictedStories;
+  }
+
+  /**
+   * @return {!Array<number>}
+   */
+  getAvailableIframeIdx() {
+    return this.iframePool_;
+  }
+
+  /**
+   * Finds adjacent iframe indices given an index.
+   * Examples of resulting adjacent arrays (x is provided index):
+   * [x] [1] [2] [] []
+   * [2] [x] [1] [] []
+   * [] [] [2] [1] [x]
+   * [1] [x]
+   * @param {number} storyIdx
+   * @param {number} maxIdx
+   * @return {!Array<number>}
+   */
+  findAdjacent(storyIdx, maxIdx) {
+    const queue = [storyIdx];
+    const adjacent = [];
+
+    const findAdjacentRecursive = () => {
+      const idx = queue.shift();
+
+      if (idx < 0 || idx > maxIdx) {
+        return;
+      }
+
+      adjacent.push(idx);
+
+      if (!queue.includes(idx + 1) && !adjacent.includes(idx + 1)) {
+        queue.push(idx + 1);
+      }
+
+      if (!queue.includes(idx - 1) && !adjacent.includes(idx - 1)) {
+        queue.push(idx - 1);
+      }
+
+      while (adjacent.length < this.iframePool_.length && queue.length > 0) {
+        findAdjacentRecursive();
+      }
+    };
+
+    findAdjacentRecursive();
+
+    return adjacent;
```

Comment:
Re: what we were discussing offline, there might be much easier implementations, like this one (I don't think it handles the extremums like you do):

```
const findAdjacent = (index, max) => {
  const adj = [];
  let cursor = index - 1;
  for (let i = 0; i < 3; i++) {
    if (cursor < 0) ++cursor;
    if (cursor > max - 1) continue;
    adj.push(cursor++);
  }
  return adj;
};

console.log(findAdjacent(0, 2));
console.log(findAdjacent(0, 10));
console.log(findAdjacent(3, 10));
console.log(findAdjacent(9, 10));
```

Reply:
Changed to iterative implementation.